File: ios/algorithms/RandomUtils.m
#import "RandomUtils.h"
#import "CryptoUtils.h"
#import <Security/Security.h>

@implementation RandomUtils

static NSString * const ALPHANUMERIC_CHARS = @"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

+ (NSString *)generateUuid {
    return [[NSUUID UUID] UUIDString];
}

+ (nullable NSString *)generateRandomKeyHex:(NSUInteger)length {
    NSData *keyData = [self generateRandomBytes:length];
    if (keyData) {
        return [CryptoUtils bytesToHex:keyData];
    }
    return nil;
}

+ (nullable NSString *)generateRandomBytesBase64:(NSUInteger)size {
    NSData *randomData = [self generateRandomBytes:size];
    if (randomData) {
        return [CryptoUtils encodeBase64:randomData];
    }
    return nil;
}

+ (nullable NSData *)generateRandomBytes:(NSUInteger)size {
    NSMutableData *randomData = [NSMutableData dataWithLength:size];
    int result = SecRandomCopyBytes(kSecRandomDefault, randomData.length, randomData.mutableBytes);
    if (result == errSecSuccess) {
        return randomData;
    } else {
        NSLog(@"Failed to generate random bytes");
        return nil;
    }
}

+ (NSString *)generateRandomAlphanumeric:(NSUInteger)length {
    NSMutableString *result = [NSMutableString stringWithCapacity:length];
    
    for (NSUInteger i = 0; i < length; i++) {
        uint32_t randomIndex = arc4random_uniform((uint32_t)[ALPHANUMERIC_CHARS length]);
        [result appendFormat:@"%C", [ALPHANUMERIC_CHARS characterAtIndex:randomIndex]];
    }
    
    return [NSString stringWithString:result];
}

+ (NSInteger)generateRandomIntMin:(NSInteger)min max:(NSInteger)max {
    if (max <= min) {
        return min;
    }
    uint32_t range = (uint32_t)(max - min);
    return min + (NSInteger)arc4random_uniform(range);
}

+ (BOOL)generateRandomBoolean {
    return arc4random_uniform(2) == 1;
}

+ (nullable NSData *)generateRandomIV:(NSUInteger)size {
    if (size == 0) {
        size = 16; // Default AES block size
    }
    return [self generateRandomBytes:size];
}

+ (nullable NSString *)generateRandomIVHex:(NSUInteger)size {
    if (size == 0) {
        size = 16; // Default AES block size
    }
    return [self generateRandomKeyHex:size];
}

+ (nullable NSData *)generateRandomSalt:(NSUInteger)size {
    if (size == 0) {
        size = 32; // Default salt size
    }
    return [self generateRandomBytes:size];
}

+ (nullable NSString *)generateRandomSaltBase64:(NSUInteger)size {
    if (size == 0) {
        size = 32; // Default salt size
    }
    return [self generateRandomBytesBase64:size];
}

@end

File: android/src/main/java/com/rocketchat/mobilecrypto/algorithms/RandomUtils.kt
package com.rocketchat.mobilecrypto.algorithms

import java.security.SecureRandom
import java.util.UUID

/**
 * Random value generation utilities
 */
object RandomUtils {

    private const val ALPHANUMERIC_CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

    /**
     * Generate a random UUID string
     * 
     * @return UUID string
     */
    fun generateUuid(): String {
        return UUID.randomUUID().toString()
    }

    /**
     * Generate a random key as hexadecimal string
     * 
     * @param length Key length in bytes
     * @return Hexadecimal key string
     */
    fun generateRandomKeyHex(length: Int): String {
        val key = ByteArray(length)
        val secureRandom = SecureRandom()
        secureRandom.nextBytes(key)
        return CryptoUtils.bytesToHex(key)
    }

    /**
     * Generate random bytes as Base64 string
     * 
     * @param size Number of bytes to generate
     * @return Base64-encoded random bytes
     */
    fun generateRandomBytesBase64(size: Int): String {
        val bytes = ByteArray(size)
        val secureRandom = SecureRandom()
        secureRandom.nextBytes(bytes)
        return CryptoUtils.encodeBase64NoWrap(bytes)
    }

    /**
     * Generate random bytes as byte array
     * 
     * @param size Number of bytes to generate
     * @return Random byte array
     */
    fun generateRandomBytes(size: Int): ByteArray {
        val bytes = ByteArray(size)
        val secureRandom = SecureRandom()
        secureRandom.nextBytes(bytes)
        return bytes
    }

    /**
     * Generate random alphanumeric string
     * 
     * @param length Length of the string to generate
     * @return Random alphanumeric string
     */
    fun generateRandomAlphanumeric(length: Int): String {
        val secureRandom = SecureRandom()
        return (1..length)
            .map { ALPHANUMERIC_CHARS[secureRandom.nextInt(ALPHANUMERIC_CHARS.length)] }
            .joinToString("")
    }

    /**
     * Generate random integer within specified range
     * 
     * @param min Minimum value (inclusive)
     * @param max Maximum value (exclusive)
     * @return Random integer
     */
    fun generateRandomInt(min: Int, max: Int): Int {
        val secureRandom = SecureRandom()
        return secureRandom.nextInt(max - min) + min
    }

    /**
     * Generate cryptographically secure random boolean
     * 
     * @return Random boolean value
     */
    fun generateRandomBoolean(): Boolean {
        val secureRandom = SecureRandom()
        return secureRandom.nextBoolean()
    }

    /**
     * Generate random IV (Initialization Vector) for AES
     * 
     * @param size IV size in bytes (default: 16 for AES)
     * @return Random IV as byte array
     */
    fun generateRandomIV(size: Int = 16): ByteArray {
        return generateRandomBytes(size)
    }

    /**
     * Generate random IV (Initialization Vector) for AES as hex string
     * 
     * @param size IV size in bytes (default: 16 for AES)
     * @return Random IV as hexadecimal string
     */
    fun generateRandomIVHex(size: Int = 16): String {
        return generateRandomKeyHex(size)
    }

    /**
     * Generate random salt for password hashing
     * 
     * @param size Salt size in bytes (default: 32)
     * @return Random salt as byte array
     */
    fun generateRandomSalt(size: Int = 32): ByteArray {
        return generateRandomBytes(size)
    }

    /**
     * Generate random salt for password hashing as Base64 string
     * 
     * @param size Salt size in bytes (default: 32)
     * @return Random salt as Base64 string
     */
    fun generateRandomSaltBase64(size: Int = 32): String {
        return generateRandomBytesBase64(size)
    }
}
